keras_retinanet/preprocessing/open_images.py

```diff
@@ -0,0 +1,201 @@
+
+            x1_abs = int(round(x1 * width))
+            x2_abs = int(round(x2 * width))
+            y1_abs = int(round(y1 * height))
+            y2_abs = int(round(y2 * height))
+
+            if y2_abs <= y1_abs:
+                print ('{} w: {} h: {}'.format(row, width, height))
+                continue
+
+            if x2_abs <= x1_abs:
+                print ('{} w: {} h: {}'.format(row, width, height))
+                continue
+
+            img_id = row['ImageID']
+
+            annotation = {'cls_id': cls_id, 'x1': x1, 'x2': x2, 'y1': y1, 'y2': y2}
+
+            if img_id in id_annotations:
+
+                annotations = id_annotations[img_id]
+                annotations['boxes'].append(annotation)
+            else:
+                id_annotations[img_id] = {'w': width, 'h': height, 'boxes': [annotation]}
+    return id_annotations
+
+
+class OpenImagesGenerator(Generator):
+    def __init__(
+            self, main_dir, subset, labels_filter,
+            **kwargs
+    ):
+        metadata_dir = os.path.join(main_dir, '2017_11')
+        fname_json = os.path.join(metadata_dir, subset, subset + '.json')
+        self.base_dir = os.path.join(main_dir, 'images', subset)
+
+        labels_dict = None if labels_filter is None else dict([(l, i) for i, l in enumerate(labels_filter)])
+
+        print ('loading {} subset'.format(subset))
+        self.id2labels_dict, cls_index_dict = get_labels(metadata_dir)
+
+        if os.path.exists(fname_json):
+            with open(fname_json, 'r') as f:
+                self.annotations = json.loads(f.read())
+        else:
+            self.annotations = generate_images_annotations_json(main_dir, metadata_dir, subset, cls_index_dict)
+            json.dump(self.annotations, open(fname_json, "w"))
+
+        if labels_dict is not None:
+            filtered_annotations = dict()
+
+            for k in self.annotations:
+                img_ann = self.annotations[k]
+
+                filtered_boxes = []
+                for ann in img_ann['boxes']:
+                    cls_id = ann['cls_id']
+                    label = self.id2labels_dict[cls_id]
+                    if label in labels_dict:
+                        ann['cls_id'] = labels_dict[label]
+                        filtered_boxes.append(ann)
+
+                if len(filtered_boxes) > 0:
+                    filtered_annotations[k] = {'w': img_ann['w'], 'h': img_ann['h'], 'boxes': filtered_boxes}
+
+            self.id2labels_dict = dict([(labels_dict[k], k) for k in labels_dict])
+            self.annotations = filtered_annotations
+
+        self.id2imageid = dict()
+        for i, k in enumerate(self.annotations):
+            self.id2imageid[i] = k
+
+        super(OpenImagesGenerator, self).__init__(**kwargs)
+
+    def size(self):
+        return len(self.annotations)
+
+    def num_classes(self):
+        return len(self.id2labels_dict)
+
+    def name_to_label(self, name):
+        raise NotImplementedError()
```

Comment:
this is easy to implement by just reversing the id2labels_dict, maybe we should have it?

Reply:
In this dataset some labels (like Bench, for ex) can have different ids since they mean different things. So does it have much sense to implement this method?

Reply:
A quick grep through the repository shows that `name_to_label` is never used outside of generators themselves. With that in mind I think it's fine to keep it unimplemented.

I'm half thinking we should remove it from the base generator API. If there are datasets where there is no 1:1 mapping between class names and ID's, the function could lead to runtime failures if someone actually tries to use it.

Reply:
This happens particulary because the labelnames in OID are things like "/m/04hgtk". And each label has a description (ex: car, person, etc...). Some labenames share the same description, but have different meaning.

Reply:
If a function in a class is never called outside the class, yes, it might be a good idea to make it internal/private. And remove it from the base generator